File: backend/app/api/routes/procurement.py
"""
Procurement API Routes
Suppliers and Purchase Orders Management
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import func, or_
from typing import List, Optional
from datetime import date, datetime
from pydantic import BaseModel

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.auth import User
from app.models.procurement import Supplier, PurchaseOrder, POItem, POStatus, SupplierStatus
from app.models.inventory import Material

router = APIRouter()

# Schemas
class SupplierCreate(BaseModel):
    supplier_code: str
    name: str
    contact_person: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    payment_terms: Optional[str] = None

class SupplierResponse(BaseModel):
    supplier_id: int
    supplier_code: str
    name: str
    rating: Optional[float] = None
    status: str
    
    class Config:
        from_attributes = True

class POItemCreate(BaseModel):
    material_id: int
    quantity: float
    unit_price: float

class POCreate(BaseModel):
    supplier_id: int
    order_date: date
    expected_delivery_date: Optional[date] = None
    items: List[POItemCreate]
    notes: Optional[str] = None

class POResponse(BaseModel):
    po_id: int
    po_number: str
    supplier_id: int
    order_date: date
    status: str
    total_amount: Optional[float] = None
    
    class Config:
        from_attributes = True

# Supplier Routes
@router.get("/suppliers", response_model=List[SupplierResponse])
async def list_suppliers(
    skip: int = 0,
    limit: int = 100,
    status: Optional[str] = None,
    search: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all suppliers with filters"""
    query = db.query(Supplier)
    
    if status:
        query = query.filter(Supplier.status == status)
    
    if search:
        query = query.filter(
            or_(
                Supplier.name.ilike(f"%{search}%"),
                Supplier.supplier_code.ilike(f"%{search}%")
            )
        )
    
    suppliers = query.offset(skip).limit(limit).all()
    return suppliers

@router.post("/suppliers", response_model=SupplierResponse, status_code=201)
async def create_supplier(
    supplier_data: SupplierCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create new supplier"""
    # Check if code exists
    if db.query(Supplier).filter(Supplier.supplier_code == supplier_data.supplier_code).first():
        raise HTTPException(status_code=400, detail="Supplier code already exists")
    
    new_supplier = Supplier(**supplier_data.dict())
    db.add(new_supplier)
    db.commit()
    db.refresh(new_supplier)
    
    return new_supplier

@router.get("/suppliers/{supplier_id}", response_model=SupplierResponse)
async def get_supplier(
    supplier_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get supplier details"""
    supplier = db.query(Supplier).filter(Supplier.supplier_id == supplier_id).first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")
    return supplier

# Purchase Order Routes
@router.get("/purchase-orders", response_model=List[POResponse])
async def list_purchase_orders(
    skip: int = 0,
    limit: int = 100,
    status: Optional[str] = None,
    supplier_id: Optional[int] = None,
    date_from: Optional[date] = None,
    date_to: Optional[date] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List purchase orders with filters"""
    query = db.query(PurchaseOrder)
    
    if status:
        query = query.filter(PurchaseOrder.status == status)
    if supplier_id:
        query = query.filter(PurchaseOrder.supplier_id == supplier_id)
    if date_from:
        query = query.filter(PurchaseOrder.order_date >= date_from)
    if date_to:
        query = query.filter(PurchaseOrder.order_date <= date_to)
    
    orders = query.order_by(PurchaseOrder.order_date.desc()).offset(skip).limit(limit).all()
    return orders

@router.post("/purchase-orders", response_model=POResponse, status_code=201)
async def create_purchase_order(
    po_data: POCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create new purchase order"""
    # Generate PO number
    po_number = f"PO-{datetime.now().strftime('%Y%m%d')}-{db.query(func.count(PurchaseOrder.po_id)).scalar() + 1:04d}"
    
    # Create PO
    new_po = PurchaseOrder(
        po_number=po_number,
        supplier_id=po_data.supplier_id,
        order_date=po_data.order_date,
        expected_delivery_date=po_data.expected_delivery_date,
        status=POStatus.DRAFT,
        notes=po_data.notes,
        created_by=current_user.user_id
    )
    
    db.add(new_po)
    db.flush()
    
    # Add PO items
    total_amount = 0
    for item_data in po_data.items:
        item_total = item_data.quantity * item_data.unit_price
        total_amount += item_total
        
        po_item = POItem(
            po_id=new_po.po_id,
            material_id=item_data.material_id,
            quantity=item_data.quantity,
            unit_price=item_data.unit_price,
            total_price=item_total
        )
        db.add(po_item)
    
    new_po.total_amount = total_amount
    new_po.net_amount = total_amount
    
    db.commit()
    db.refresh(new_po)
    
    return new_po

@router.get("/purchase-orders/{po_id}")
async def get_purchase_order(
    po_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get PO details with items"""
    po = db.query(PurchaseOrder).filter(PurchaseOrder.po_id == po_id).first()
    if not po:
        raise HTTPException(status_code=404, detail="Purchase order not found")
    
    return {
        "po": po,
        "items": po.items,
        "supplier": po.supplier
    }

@router.put("/purchase-orders/{po_id}/approve")
async def approve_purchase_order(
    po_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Approve purchase order"""
    po = db.query(PurchaseOrder).filter(PurchaseOrder.po_id == po_id).first()
    if not po:
        raise HTTPException(status_code=404, detail="Purchase order not found")
    
    if po.status != POStatus.DRAFT and po.status != POStatus.PENDING_APPROVAL:
        raise HTTPException(status_code=400, detail="PO cannot be approved in current status")
    
    po.status = POStatus.APPROVED
    po.approved_by = current_user.user_id
    po.approved_at = datetime.utcnow()
    
    db.commit()
    
    return {"message": "Purchase order approved successfully", "po_number": po.po_number}

@router.put("/purchase-orders/{po_id}/receive")
async def receive_purchase_order(
    po_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Mark PO as received"""
    po = db.query(PurchaseOrder).filter(PurchaseOrder.po_id == po_id).first()
    if not po:
        raise HTTPException(status_code=404, detail="Purchase order not found")
    
    po.status = POStatus.RECEIVED
    po.actual_delivery_date = date.today()
    
    db.commit()
    
    return {"message": "Purchase order marked as received", "po_number": po.po_number}

@router.get("/pending-approvals")
async def get_pending_approvals(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get POs pending approval"""
    pending_pos = db.query(PurchaseOrder).filter(
        or_(
            PurchaseOrder.status == POStatus.PENDING_APPROVAL,
            PurchaseOrder.status == POStatus.DRAFT
        )
    ).order_by(PurchaseOrder.order_date.desc()).all()
    
    return {"count": len(pending_pos), "purchase_orders": pending_pos}

@router.post("/delivery-update")
async def update_delivery_status(
    po_id: int,
    actual_delivery_date: date,
    notes: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update delivery status"""
    po = db.query(PurchaseOrder).filter(PurchaseOrder.po_id == po_id).first()
    if not po:
        raise HTTPException(status_code=404, detail="Purchase order not found")
    
    po.actual_delivery_date = actual_delivery_date
    if notes:
        po.notes = (po.notes or "") + f"\nDelivery update: {notes}"
    
    db.commit()
    
    return {"message": "Delivery status updated", "po_number": po.po_number}

File: backend-node/routes/procurement.js
const express = require('express');
const router = express.Router();
const { Op } = require('sequelize');
const { Supplier, PurchaseOrder, POItem, Material, sequelize } = require('../models');
const { authenticateToken } = require('../middleware/auth');

// 1. List suppliers
router.get('/suppliers', authenticateToken, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 100;
    const offset = parseInt(req.query.skip) || 0;
    const { status, search } = req.query;

    const whereClause = {};
    if (status) {
      whereClause.status = status;
    }
    if (search) {
      whereClause[Op.or] = [
        { name: { [Op.like]: `%${search}%` } },
        { supplier_code: { [Op.like]: `%${search}%` } }
      ];
    }

    const suppliers = await Supplier.findAll({
      where: whereClause,
      limit,
      offset
    });

    return res.json(suppliers);
  } catch (error) {
    console.error('List Suppliers Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 2. Create supplier
router.post('/suppliers', authenticateToken, async (req, res) => {
  try {
    const { supplier_code, name, contact_person, email, phone, address, city, payment_terms } = req.body;

    const existing = await Supplier.findOne({ where: { supplier_code } });
    if (existing) {
      return res.status(400).json({ detail: 'Supplier code already exists' });
    }

    const newSupplier = await Supplier.create({
      supplier_code,
      name,
      contact_person,
      email,
      phone,
      address,
      city,
      payment_terms
    });

    return res.status(201).json(newSupplier);
  } catch (error) {
    console.error('Create Supplier Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 3. Get supplier details
router.get('/suppliers/:supplier_id', authenticateToken, async (req, res) => {
  try {
    const supplier = await Supplier.findByPk(parseInt(req.params.supplier_id));
    if (!supplier) {
      return res.status(404).json({ detail: 'Supplier not found' });
    }
    return res.json(supplier);
  } catch (error) {
    console.error('Get Supplier Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 4. List purchase orders
router.get('/purchase-orders', authenticateToken, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 100;
    const offset = parseInt(req.query.skip) || 0;
    const { status, supplier_id, date_from, date_to } = req.query;

    const whereClause = {};
    if (status) {
      whereClause.status = status;
    }
    if (supplier_id) {
      whereClause.supplier_id = parseInt(supplier_id);
    }
    if (date_from) {
      whereClause.order_date = { ...whereClause.order_date, [Op.gte]: date_from };
    }
    if (date_to) {
      whereClause.order_date = { ...whereClause.order_date, [Op.lte]: date_to };
    }

    const orders = await PurchaseOrder.findAll({
      where: whereClause,
      order: [['order_date', 'DESC']],
      limit,
      offset
    });

    return res.json(orders);
  } catch (error) {
    console.error('List POs Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 5. Create new purchase order
router.post('/purchase-orders', authenticateToken, async (req, res) => {
  const t = await sequelize.transaction();
  try {
    const { supplier_id, order_date, expected_delivery_date, items, notes } = req.body;

    // Generate PO number (format: PO-YYYYMMDD-XXXX)
    const todayStr = new Date().toISOString().slice(0, 10).replace(/-/g, '');
    const count = await PurchaseOrder.count({ transaction: t });
    const po_number = `PO-${todayStr}-${(count + 1).toString().padStart(4, '0')}`;

    const new_po = await PurchaseOrder.create({
      po_number,
      supplier_id,
      order_date,
      expected_delivery_date,
      status: 'draft',
      notes,
      created_by: req.user.user_id
    }, { transaction: t });

    let total_amount = 0;
    for (const item of items) {
      const item_total = item.quantity * item.unit_price;
      total_amount += item_total;

      await POItem.create({
        po_id: new_po.po_id,
        material_id: item.material_id,
        quantity: item.quantity,
        unit_price: item.unit_price,
        total_price: item_total
      }, { transaction: t });
    }

    new_po.total_amount = total_amount;
    new_po.net_amount = total_amount;
    await new_po.save({ transaction: t });

    await t.commit();

    return res.status(201).json(new_po);

  } catch (error) {
    await t.rollback();
    console.error('Create PO Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 6. Get purchase order details
router.get('/purchase-orders/:po_id', authenticateToken, async (req, res) => {
  try {
    const po_id = parseInt(req.params.po_id);
    const po = await PurchaseOrder.findByPk(po_id, {
      include: [
        { model: POItem, as: 'items', include: [{ model: Material, as: 'material' }] },
        { model: Supplier, as: 'supplier' }
      ]
    });

    if (!po) {
      return res.status(404).json({ detail: 'Purchase order not found' });
    }

    return res.json({
      po,
      items: po.items,
      supplier: po.supplier
    });
  } catch (error) {
    console.error('Get PO Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 7. Approve purchase order
router.put('/purchase-orders/:po_id/approve', authenticateToken, async (req, res) => {
  try {
    const po_id = parseInt(req.params.po_id);
    const po = await PurchaseOrder.findByPk(po_id);
    if (!po) {
      return res.status(404).json({ detail: 'Purchase order not found' });
    }

    if (po.status !== 'draft' && po.status !== 'pending_approval') {
      return res.status(400).json({ detail: 'PO cannot be approved in current status' });
    }

    po.status = 'approved';
    po.approved_by = req.user.user_id;
    po.approved_at = new Date();
    await po.save();

    return res.json({ message: 'Purchase order approved successfully', po_number: po.po_number });
  } catch (error) {
    console.error('Approve PO Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 8. Mark PO as received
router.put('/purchase-orders/:po_id/receive', authenticateToken, async (req, res) => {
  try {
    const po_id = parseInt(req.params.po_id);
    const po = await PurchaseOrder.findByPk(po_id);
    if (!po) {
      return res.status(404).json({ detail: 'Purchase order not found' });
    }

    po.status = 'received';
    po.actual_delivery_date = new Date();
    await po.save();

    return res.json({ message: 'Purchase order marked as received', po_number: po.po_number });
  } catch (error) {
    console.error('Receive PO Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 9. Get POs pending approval
router.get('/pending-approvals', authenticateToken, async (req, res) => {
  try {
    const pending_pos = await PurchaseOrder.findAll({
      where: {
        status: { [Op.in]: ['pending_approval', 'draft'] }
      },
      order: [['order_date', 'DESC']]
    });

    return res.json({ count: pending_pos.length, purchase_orders: pending_pos });
  } catch (error) {
    console.error('Get Pending Approvals Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

// 10. Update delivery status
router.post('/delivery-update', authenticateToken, async (req, res) => {
  try {
    const { po_id, actual_delivery_date, notes } = req.body;

    const po = await PurchaseOrder.findByPk(po_id);
    if (!po) {
      return res.status(404).json({ detail: 'Purchase order not found' });
    }

    po.actual_delivery_date = actual_delivery_date;
    if (notes) {
      po.notes = (po.notes || '') + `\nDelivery update: ${notes}`;
    }
    await po.save();

    return res.json({ message: 'Delivery status updated', po_number: po.po_number });
  } catch (error) {
    console.error('Delivery Update Error:', error);
    return res.status(500).json({ detail: 'Internal server error' });
  }
});

module.exports = router;
